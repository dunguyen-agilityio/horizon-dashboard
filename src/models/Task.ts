import { TUser } from './User';

export const enum LABEL {
  PENDING = 'Pending',
  UPDATES = 'Updates',
  DONE = 'Done',
  ERRORS = 'Errors',
}

export const enum STATUS {
  BACKLOG = 'backlog',
  IN_PROGRESS = 'in-progress',
  DONE = 'done',
}

export class Task {
  id!: string;
  title!: string;
  description!: string;
  createdAt!: Date | string;
  updatedAt!: Date | string;
  dueDate?: string;
  startedDate?: string;
  status!: STATUS;
  labels!: LABEL[];
  assignees?: TUser[];
  cover?: string;

  constructor(task?: Task) {
    const {
      title = '',
      description = '',
      createdAt,
      updatedAt,
      dueDate,
      startedDate,
      assignees = [],
      labels = [],
    } = task || {};

    Object.assign(this, {
      ...task,
      title,
      description,
      assignees,
      labels: labels.length ? labels : [LABEL.PENDING],
      dueDate: dueDate ?? '',
      createdAt: createdAt ? new Date(createdAt) : undefined,
      updatedAt: updatedAt ? new Date(updatedAt) : undefined,
      startedDate: startedDate ?? '',
    });
  }
}
