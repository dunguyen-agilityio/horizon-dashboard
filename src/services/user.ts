// Models
import { TUser } from '@/models/User';

// Services
import { apiClient } from './api';

// Constants
import { API_ENTITY } from '@/constants';

export const getTeamMember = () =>
  apiClient.get<TUser[]>(
    `${API_ENTITY.USERS}?limit=3&populate[0]=avatar&populate[1]=role`,
  );
