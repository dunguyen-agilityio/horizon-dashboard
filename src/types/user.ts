// Model
import { TImage } from '@/models/Image';
import { TUser } from '@/models/User';

export type Role = { name: string };

export type UserExtend = Omit<TUser, 'avatar' | 'fullName'> & {
  avatar?: TImage;
};
