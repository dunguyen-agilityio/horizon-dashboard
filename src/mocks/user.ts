import { TImage } from '@/models/Image';
import { TUser } from '@/models/User';
import { StrapiResponse } from '@/types/strapi';
import { UserResponse1 } from '@/utils/nft';

export const MOCK_AVATAR: TImage = {
  hash: 'image_de1e29b0b7',
  url: 'https://res.cloudinary.com/dn5vw6fjp/image/upload/v1726630608/image_de1e29b0b7.png',
};

export const MOCK_USERS: TUser[] = [
  {
    id: '1',
    username: 'Adela Parkson',
    firstName: 'User 1',
    email: 'user1@gmail.com',
    role: { name: 'Creative Director' },
    createdAt: new Date('2024-08-17T07:25:04.188Z'),
    rating: 29,
    avatar: MOCK_AVATAR,
  },
  {
    id: '2',
    username: 'user2',
    firstName: 'User 2',
    email: 'user2@gmail.com',
    role: { name: 'Product Designer' },
    createdAt: new Date('2024-08-13T07:25:04.188Z'),
    rating: 72,
    avatar: MOCK_AVATAR,
  },
  {
    id: '3',
    username: 'user3',
    firstName: 'User 3',
    email: 'user3@gmail.com',
    role: { name: 'Product Designer' },
    createdAt: new Date('2024-01-13T07:25:04.188Z'),
    rating: 89,
    avatar: MOCK_AVATAR,
  },
  {
    id: '4',
    username: 'user4',
    firstName: 'User 4',
    email: 'user4@gmail.com',
    role: { name: 'Junior Graphic Designer' },
    createdAt: new Date('2024-08-18T07:25:04.188Z'),
    rating: 52,
    avatar: MOCK_AVATAR,
  },
];

export const MOCK_ROLE_RESPONSE: StrapiResponse<{ id: string; name: string }> =
  {
    id: '1',
    attributes: { name: 'Developer' },
  };

export const MOCK_IMAGE_RESPONSE: StrapiResponse<TImage> = {
  id: '1',
  attributes: {
    hash: 'image_de1e29b0b7',
    url: 'https://res.cloudinary.com/dn5vw6fjp/image/upload/v1726630608/image_de1e29b0b7.png',
  },
};

export const MOCK_USERS_RESPONSE: StrapiResponse<UserResponse1>[] =
  MOCK_USERS.map(({ id, ...attributes }) => ({
    attributes: {
      ...attributes,
      avatar: { data: MOCK_IMAGE_RESPONSE },
      role: MOCK_ROLE_RESPONSE,
    },
    id,
  }));
