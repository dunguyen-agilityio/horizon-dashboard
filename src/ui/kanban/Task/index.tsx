// Components
import Image from 'next/image';
import Pen from '@/icons/Pen';
import Text from '@/components/Text';
import { Chip } from '@nextui-org/chip';
import Button from '@/components/Button';
import { Avatar, AvatarGroup } from '@nextui-org/avatar';

// Types
import { TEXT_SIZE, TEXT_VARIANT } from '@/types/text';

// Models
import { Task as TaskModel } from '@/models/Task';

// Utils
import { getColorByLabel } from '@/utils/task';

export const Task = ({
  title,
  description,
  labels,
  cover,
  assignees,
}: Pick<
  TaskModel,
  'title' | 'description' | 'labels' | 'cover' | 'assignees'
>) => (
  <div className="flex flex-col gap-[10px] pt-[18px] pl-[23px] pb-[23px] pr-[21px] rounded-[15px] bg-white dark:bg-indigo-light max-w-[466px]">
    <div className="flex justify-between">
      <Text size={TEXT_SIZE.lg} as="h3" className="max-w-[90%]">
        {title}
      </Text>
      <Button size="fit" aria-label="edit-button" className="bg-transparent">
        <Pen />
      </Button>
    </div>
    {cover && (
      <div className="relative min-h-[285px] min-w-[308px]">
        <Image src={cover} alt="cover" fill />
      </div>
    )}
    <Text as="p" variant={TEXT_VARIANT.SECONDARY}>
      {description}
    </Text>
    <div className="flex justify-between mt-5 items-center">
      <AvatarGroup isBordered max={3}>
        {assignees?.map(({ id, avatar, fullName }) => (
          <Avatar
            key={id}
            src={avatar}
            ImgComponent={Image}
            alt={fullName}
            imgProps={{ width: 20, height: 20 }}
            isBordered
          />
        ))}
      </AvatarGroup>
      <Chip
        color={getColorByLabel(labels[0])}
        className="text-white uppercase text-xs leading-[15px] rounded-[10px] py-[6.5px] px-[25px]"
      >
        {labels[0]}
      </Chip>
    </div>
  </div>
);
