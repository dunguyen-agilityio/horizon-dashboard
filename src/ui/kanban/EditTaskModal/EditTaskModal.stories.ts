import type { Meta, StoryObj } from '@storybook/react';
import { fn } from '@storybook/test';

// Components
import { EditTaskModal } from '.';

// Models
import { LABEL } from '@/models/Task';

// Mocks
import { MOCK_USERS } from '@/mocks/user';

const meta = {
  title: 'Components/EditTaskModal',
  component: EditTaskModal,
  tags: ['autodocs'],
} satisfies Meta<typeof EditTaskModal>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    isOpen: true,
    title: 'Option to use local/server version feature',
    labels: [LABEL.DONE],
    description:
      'It usually displays this message when you close an unsaved page when you do it on purpose',
    onOpenChange: fn(),
    assignMembers: MOCK_USERS,
  },
};
