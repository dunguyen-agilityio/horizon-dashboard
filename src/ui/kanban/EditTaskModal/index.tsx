import dynamic from 'next/dynamic';
import { useMemo } from 'react';
import { useForm, Controller } from 'react-hook-form';
import { parseAbsoluteToLocal, ZonedDateTime } from '@internationalized/date';
import {
  EditorState,
  convertToRaw,
  convertFromRaw,
  RawDraftContentState,
} from 'draft-js';
import 'react-draft-wysiwyg/dist/react-draft-wysiwyg.css';

// Components
import {
  Modal,
  ModalContent,
  ModalHeader,
  ModalBody,
  ModalFooter,
  Button,
  Input,
  Select,
  SelectItem,
  DatePicker,
  SharedSelection,
} from '@nextui-org/react';

import { BoxIcon, Text } from '@/components';

// hooks
import { useTask } from '@/hooks/useTask';
import useToast from '@/contexts/toast';

// Types
import { DateTimeParts } from '@/types/date';
import { TEXT_SIZE } from '@/types/text';

// Models
import { LABEL, STATUS } from '@/models/Task';
import { TUser, User } from '@/models/User';

// Mocks
import { Note } from '@/icons';
import { MOCK_USERS } from '@/mocks/user';
import { UPDATE_DUE_DATE_ERROR, UPDATE_START_DATE_ERROR } from '@/mocks/toast';

// utils
import { convertToUTCString } from '@/utils/format';
import { getUpdatedAssignMembers } from '@/utils/task';
import { validateDates } from '@/utils/validation';

interface IEditTaskModalProps {
  id: string;
  isOpen: boolean;
  title: string;
  status: STATUS;
  description?: RawDraftContentState;
  labels?: LABEL[];
  assignMembers?: TUser[];
  startDateTask: string;
  dueDateTask: string;
  onOpenChange: () => void;
  onClose: () => void;
}

const labelValues = [
  { value: 'Pending' },
  { value: 'Done' },
  { value: 'Errors' },
  { value: 'Updates' },
];

const statusValues = [
  { value: 'backlog' },
  { value: 'in-progress' },
  { value: 'done' },
];

const selectStyle = {
  base: 'bg-white dark:bg-indigo',
  label:
    'group-data-[filled=true]:dark:text-white group-data-[filled=true]:text-primary',
  listbox: 'text-primary dark:text-secondary',
  value:
    'group-data-[has-value=true]:text-primary group-data-[has-value=true]:dark:text-white capitalize',
  listboxWrapper: 'bg-white dark:bg-indigo',
  popoverContent: 'bg-white dark:bg-indigo',
};

interface IEditTaskProps {
  title: string;
  status: STATUS[];
  description?: EditorState;
  selectedMembers: string[];
  selectedLabels: LABEL[];
  startDate: DateTimeParts | null;
  dueDate: DateTimeParts | null;
}

const Editor = dynamic(
  () => import('react-draft-wysiwyg').then((mod) => mod.Editor),
  {
    ssr: false,
  },
);

export const EditTaskModal = ({
  id,
  isOpen,
  title,
  status,
  description,
  labels = [],
  assignMembers = [],
  onOpenChange,
  onClose,
  startDateTask,
  dueDateTask,
}: IEditTaskModalProps) => {
  const editorState = description
    ? EditorState.createWithContent(convertFromRaw(description))
    : EditorState.createEmpty();

  const initialValue = useMemo(
    () => ({
      title: title || '',
      status: [status],
      description: editorState,
      startDate: startDateTask ? parseAbsoluteToLocal(startDateTask) : null,
      dueDate: dueDateTask ? parseAbsoluteToLocal(dueDateTask) : null,
      selectedLabels: labels.map((label) => label),
      selectedMembers: assignMembers.map(({ username }) => username),
    }),
    [
      title,
      status,
      editorState,
      startDateTask,
      dueDateTask,
      labels,
      assignMembers,
    ],
  );

  const { showToast } = useToast();

  const { control, handleSubmit, reset, formState, getValues } = useForm({
    defaultValues: initialValue,
  });

  const handleLabelChange = (
    keys: SharedSelection,
    onChange: (value: LABEL[]) => void,
  ) => {
    const selected = Array.from(keys).map((key) => key as LABEL);
    onChange(selected);
  };

  const handleStatusChange = (
    keys: SharedSelection,
    onChange: (value: STATUS[]) => void,
  ) => {
    const selected = Array.from(keys).map((key) => key as STATUS);
    onChange(selected);
  };

  const handleMembersChange = (
    keys: SharedSelection,
    onChange: (value: string[]) => void,
  ) => {
    const selected = Array.from(keys).map((key) => key as string);
    onChange(selected);
  };

  const { updateTask, removeTask } = useTask(status);

  const onSubmit = async (editTask: IEditTaskProps) => {
    const {
      title,
      status,
      description,
      selectedLabels,
      startDate,
      dueDate,
      selectedMembers,
    } = editTask;
    const updatedAssignMembers = getUpdatedAssignMembers(
      selectedMembers,
    ) as User[];

    const descriptionEditor = description
      ? convertToRaw(description.getCurrentContent())
      : null;

    await updateTask(
      id,
      title,
      status[0],
      descriptionEditor,
      selectedLabels,
      updatedAssignMembers,
      convertToUTCString(startDate),
      convertToUTCString(dueDate),
    );
    onClose();
  };

  const handleDelete = () => {
    removeTask(id);
    onClose();
  };

  const handleOnClose = () => {
    reset(initialValue);
    onClose();
  };

  const handleChangeStartDate = (
    date: ZonedDateTime,
    onChange: (value: ZonedDateTime | null) => void,
  ) => {
    if (!validateDates(date, getValues('dueDate'))) {
      showToast(UPDATE_START_DATE_ERROR);
      return onChange(null);
    }
    return onChange(date);
  };

  const handleChangeDueDate = (
    date: ZonedDateTime,
    onChange: (value: ZonedDateTime | null) => void,
  ) => {
    if (!validateDates(getValues('startDate'), date)) {
      showToast(UPDATE_DUE_DATE_ERROR);
      return onChange(null);
    }
    return onChange(date);
  };

  return (
    <Modal
      size="3xl"
      isOpen={isOpen}
      onOpenChange={onOpenChange}
      onClose={handleOnClose}
      placement="center"
      scrollBehavior="outside"
      classNames={{
        base: 'bg-white dark:bg-indigo pt-4 pb-4 min-h-[480px] m-2',
        closeButton: 'm-3 p-3',
      }}
    >
      <ModalContent>
        <form onSubmit={handleSubmit(onSubmit)}>
          <ModalHeader className="flex flex-col gap-1">
            <Controller
              name="title"
              control={control}
              render={({ field }) => (
                <Input
                  className="bg-transparent w-5/6"
                  classNames={{
                    input:
                      'text-lg font-[700] group-data-[has-value=true]:text-primary group-data-[has-value=true]:dark:text-white group-data-[has-value=true]:shadow-none ',
                    inputWrapper: 'bg-transparent shadow-none',
                  }}
                  type="text"
                  as="h2"
                  {...field}
                />
              )}
            />
          </ModalHeader>
          <ModalBody>
            <div className="flex flex-col justify-between lg:flex-row gap-4 px-3">
              <div className="flex flex-col gap-4 w-full">
                <div className="flex items-center gap-2">
                  <Text className="font-bold">Description</Text>
                  <BoxIcon icon={<Note />} customClass="fill-secondary" />
                </div>
                <div className="bg-gray dark:bg-indigo-light rounded-md editor-container w-[300px]  ">
                  <Controller
                    name="description"
                    control={control}
                    render={({ field }) => (
                      <Editor
                        toolbarClassName="editor-toolbar"
                        editorClassName="editor-description"
                        editorState={field.value}
                        onEditorStateChange={(editorState) => {
                          // setDescriptionEditorState(editorState);
                          field.onChange(editorState);
                        }}
                        toolbar={{
                          options: ['inline', 'history'],
                          inline: {
                            options: [
                              'bold',
                              'italic',
                              'underline',
                              'strikethrough',
                            ],
                          },
                          list: { options: ['unordered', 'ordered'] },
                        }}
                        placeholder="Write something here..."
                      />
                    )}
                  />
                </div>
              </div>
              <div>
                <div className="flex flex-row flex-wrap md:flex-nowrap md:flex-col lg:flex-col gap-2 md:justify-center min-w-[250px]">
                  <Text size={TEXT_SIZE.md} className="font-bold pb-1">
                    Add to card
                  </Text>
                  <Controller
                    name="status"
                    control={control}
                    render={({ field }) => (
                      <Select
                        aria-label="select-status"
                        label="Status"
                        defaultSelectedKeys={field.value}
                        selectedKeys={field.value}
                        onSelectionChange={(key: SharedSelection) =>
                          handleStatusChange(key, field.onChange)
                        }
                        classNames={selectStyle}
                      >
                        {statusValues.map((status) => (
                          <SelectItem key={status.value} className="capitalize">
                            {status.value}
                          </SelectItem>
                        ))}
                      </Select>
                    )}
                  />
                  <Controller
                    name="selectedLabels"
                    control={control}
                    render={({ field }) => (
                      <Select
                        aria-label="select-label"
                        label="Label"
                        selectedKeys={field.value}
                        onSelectionChange={(key: SharedSelection) =>
                          handleLabelChange(key, field.onChange)
                        }
                        classNames={selectStyle}
                      >
                        {labelValues.map((label) => (
                          <SelectItem key={label.value}>
                            {label.value}
                          </SelectItem>
                        ))}
                      </Select>
                    )}
                  />
                  <Controller
                    name="selectedMembers"
                    control={control}
                    render={({ field }) => (
                      <Select
                        aria-label="select-members"
                        label="Assign members"
                        selectionMode="multiple"
                        selectedKeys={field.value}
                        onSelectionChange={(key: SharedSelection) =>
                          handleMembersChange(key, field.onChange)
                        }
                        classNames={selectStyle}
                      >
                        {MOCK_USERS.map(({ username }) => (
                          <SelectItem key={username}>{username}</SelectItem>
                        ))}
                      </Select>
                    )}
                  />
                  <Controller
                    name="startDate"
                    control={control}
                    render={({ field }) => (
                      <DatePicker
                        aria-label="start-date-picker"
                        label="Start date"
                        hideTimeZone
                        showMonthAndYearPickers
                        granularity="minute"
                        value={field.value}
                        onChange={(date) =>
                          handleChangeStartDate(date, field.onChange)
                        }
                      />
                    )}
                  />
                  <Controller
                    name="dueDate"
                    control={control}
                    render={({ field }) => (
                      <DatePicker
                        aria-label="due-date-picker"
                        label="Due date"
                        showMonthAndYearPickers
                        hideTimeZone
                        granularity="minute"
                        value={field.value}
                        onChange={(date) =>
                          handleChangeDueDate(date, field.onChange)
                        }
                      />
                    )}
                  />
                </div>

                <div className="flex flex-col mt-4">
                  <Text size={TEXT_SIZE.md} className="font-bold pb-2">
                    Actions
                  </Text>
                  <Button
                    onClick={handleDelete}
                    className="w-full justify-start bg-red-750 text-white pl-3 pr-3"
                  >
                    Remove task
                  </Button>
                </div>
              </div>
            </div>
          </ModalBody>
          <ModalFooter>
            <Button
              color="primary"
              type="submit"
              isDisabled={!formState.isDirty}
            >
              Update
            </Button>
          </ModalFooter>
        </form>
      </ModalContent>
    </Modal>
  );
};

export default EditTaskModal;
