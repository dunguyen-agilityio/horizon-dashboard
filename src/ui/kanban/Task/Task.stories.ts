import type { Meta, StoryObj } from '@storybook/react';

import { Task } from '.';

import { MOCK_TASKS } from '@/mocks/task';

const mockTask = MOCK_TASKS[0];
const mockCover = '/default-task-cover.webp';

const meta = {
  title: 'Components/Task',
  component: Task,
  tags: ['autodocs'],
} satisfies Meta<typeof Task>;

export default meta;
type Story = StoryObj<typeof meta>;

export const TaskMissCover: Story = {
  args: { ...mockTask, cover: undefined },
};

export const TaskWithCover: Story = {
  args: { ...mockTask, cover: mockCover },
};
