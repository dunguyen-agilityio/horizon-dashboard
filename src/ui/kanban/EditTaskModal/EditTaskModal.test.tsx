import { render } from '@testing-library/react';

// Components
import EditTaskModal from '.';

// Models
import { LABEL } from '@/models/Task';

// Mocks
import { MOCK_USERS } from '@/mocks/user';

const titleModal = 'Option to use local/server version feature';
const descriptionModal =
  "It usually displays this message when you close an unsaved page when you do it on purpose, and it's getting frustrated to see this every time.";

const onOpenChangeMock = jest.fn();

describe('EditTaskModal tests', () => {
  it('Should match snapshot', () => {
    const { container } = render(
      <EditTaskModal
        isOpen={true}
        title={titleModal}
        labels={[LABEL.DONE]}
        description={descriptionModal}
        onOpenChange={onOpenChangeMock}
        assignMembers={MOCK_USERS}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('Should render correct title and description of Task', () => {
    const { getByText } = render(
      <EditTaskModal
        isOpen
        title={titleModal}
        labels={[LABEL.DONE]}
        description={descriptionModal}
        onOpenChange={onOpenChangeMock}
        assignMembers={MOCK_USERS}
      />,
    );
    expect(getByText(titleModal)).toBeInTheDocument();
    expect(getByText(descriptionModal)).toBeInTheDocument();
  });
});
