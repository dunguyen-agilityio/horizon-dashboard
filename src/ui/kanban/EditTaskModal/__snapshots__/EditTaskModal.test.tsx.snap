// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditTaskModal tests Should match snapshot 1`] = `
<div
  aria-hidden="true"
/>
`;
